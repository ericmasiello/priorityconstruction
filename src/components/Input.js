import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import tinyColor from 'tinycolor2';
import { COLORS, BODY_WEIGHTS } from '../styles/vars';
import * as CustomPropTypes from '../propTypes';

const RadioMark = styled.span`
  justify-content: center;
  border-radius: 50%;
  width: 20px;
  height: 20px;
  border: 1px solid ${COLORS.border};
  position: relative;
  ${({ error }) => error && `border-color: ${COLORS.error};`};
`;

const LabelText = styled.span`
  margin-left: 0.5rem;
`;

export const Radio = ({ children, error, className, ...rest }) => (
  <label className={className}>
    <input {...rest} />
    <RadioMark error={error} />
    {children && <LabelText>{children}</LabelText>}
  </label>
);

Radio.propTypes = {
  tag: CustomPropTypes.Tag,
  className: PropTypes.string,
  children: PropTypes.node,
  error: PropTypes.bool,
};

Radio.defaultProps = {
  tag: 'input',
};

Radio.displayName = 'Radio';

const StyledRadio = styled(Radio)`
  position: relative;
  display: inline-flex;
  align-items: center;
  overflow: hidden;

  input {
    position: absolute;
    left: -9999px;
  }

  input:checked + ${RadioMark} {
    &::before {
      content: '';
      background-color: ${tinyColor(COLORS.border)
        .darken(40)
        .toRgbString()};
      border-radius: 50%;
      width: 10px;
      height: 10px;
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translateX(-50%) translateY(-50%);
      ${({ error }) => error && `background-color: ${COLORS.error};`};
    }
  }
`;

export const Input = ({ tag: Tag, ...rest }) => <Tag {...rest} />;

Input.propTypes = {
  tag: CustomPropTypes.Tag,
};

Input.defaultProps = {
  tag: 'input',
};

Input.displayName = 'Input';

const StyledInput = styled(Input)`
  border: 1px solid ${COLORS.border};
  font-weight: ${BODY_WEIGHTS.medium};
  background-color: ${COLORS.bg};
  padding: 0.5rem;

  ${({ error }) => error && `border-color: ${COLORS.error}`};
`;

const Component = props => {
  switch (props.type) {
    case 'radio':
      return <StyledRadio {...props} />;
    default:
      return <StyledInput {...props} />;
  }
};

Component.displayName = 'Input';

export default styled(Component)``;
