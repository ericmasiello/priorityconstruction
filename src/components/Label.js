import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export const Label = (props) => {
  const { tag: Tag, ...rest } = props;
  return (
    <Tag {...rest} />
  );
};

Label.propTypes = {
  tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
};

Label.defaultProps = {
  tag: 'label',
};

Label.displayName = 'Label';

export default styled(Label)`
  display: block;
`;
