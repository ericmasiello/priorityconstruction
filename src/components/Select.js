import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { COLORS, BODY_WEIGHTS } from '../styles/vars';

export const SelectElement = styled.select`
  position: relative;
  z-index: 2;
  padding: 0.5rem 1.25rem 0.5rem 0.5rem;
  appearance: none;
  background-color: transparent;
  border: none;
`;

export const Select = ({ tag: Tag, children, ...rest }) => (
  <Tag {...rest}>
    <SelectElement>
      {children}
    </SelectElement>
  </Tag>
);

Select.propTypes = {
  tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  children: PropTypes.node,
};

Select.defaultProps = {
  tag: 'span',
};

Select.displayName = 'Select';

export default styled(Select)`
  position: relative;
  display: inline-block;
  border: 1px solid ${COLORS.border};
  font-weight: ${BODY_WEIGHTS.medium};
  background-color: ${COLORS.bg};

  &::after {
    content: '';
    position: absolute;
    top: 50%;
    right: 0.5rem;
    transform: translateY(-50%);
    width: 0; 
    height: 0; 
    border-left: 3px solid transparent;
    border-right: 3px solid transparent;
    border-top: 4px solid ${COLORS.base};
  }
`;
