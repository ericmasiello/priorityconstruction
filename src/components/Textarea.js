import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { COLORS, BODY_WEIGHTS } from '../styles/vars';

export const Textarea = ({ tag: Tag, error, ...rest }) => <Tag {...rest} />;

Textarea.propTypes = {
  tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  error: PropTypes.bool,
};

Textarea.defaultProps = {
  tag: 'textarea',
};

Textarea.displayName = 'Textarea';

export default styled(Textarea)`
  border: 1px solid ${COLORS.border};
  font-weight: ${BODY_WEIGHTS.medium};
  background-color: ${COLORS.bg};
  padding: 0.5rem;
  ${({ error }) => error && `border-color: ${COLORS.error}`};
`;
