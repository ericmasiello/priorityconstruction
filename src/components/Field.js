import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Label from './Label';
import Input from './Input';
import Textarea from './Textarea';
import Select from './Select';

const isLabel = (child) => child.type.displayName === Label.displayName;
const isFormField = (child) => {
  if (child.type.displayName === Input.displayName
    || child.type.displayName === Textarea.displayName
    || child.type.displayName === Select.displayName) {
    return true;
  }
  return false;
};

const applyNameAs = (child, nameAs) => {
  if (!nameAs) {
    return child;
  }

  if (isLabel(child)) {
    return React.cloneElement(child, Object.assign({
      htmlFor: nameAs,
    }, child.props));
  }

  if (isFormField(child)) {
    return React.cloneElement(child, Object.assign({
      id: nameAs,
      name: nameAs,
    }, child.props));
  }
  return child;
}

export const Field = (props) => {
  const {
    tag: Tag,
    nameAs,
    children,
    stack,
    ...rest
  } = props;
  return (
    <Tag {...rest}>
      {React.Children.map(children, (child) => applyNameAs(child, nameAs))}
    </Tag>
  );
};

Field.propTypes = {
  tag: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  stack: PropTypes.bool,
  children: PropTypes.node,
  nameAs: PropTypes.string,
};

Field.defaultProps = {
  tag: 'div',
};

Field.displayName = 'Field';

export default styled(Field)`
  display: flex;
  align-items: baseline;
  ${(props) => props.stack && 'flex-direction: column;'}
  margin-bottom: 1rem;

  ${Label} {
    ${(props) => !props.stack && `
      flex: 0 1 100px;
      max-width: 200px;
    `}
  }
  ${Input},
  ${Textarea},
  ${Select} {
    width: 100%;
  }
`;
