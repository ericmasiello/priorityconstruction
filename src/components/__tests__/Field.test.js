import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Field, { Field as Raw } from '../Field';
import Label from '../Label';
import Input from '../Input';
import Textarea from '../Textarea';
import Select from '../Select';

it('should render', () => {
  const component = renderer.create(<Field />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render as stack', () => {
  const component = renderer.create(<Field stack />);

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render children', () => {
  const wrapper = shallow(<Raw>Test</Raw>);

  expect(wrapper.text()).toBe('Test');
});

it('should render additional props', () => {
  const wrapper = shallow(<Raw data-test="my-attribute" />);

  expect(wrapper.find('[data-test="my-attribute"]')).toHaveLength(1);
});

it('should display additional classes', () => {
  const wrapper = shallow(<Raw className="my-class" />);

  expect(wrapper.hasClass('my-class')).toBe(true);
});

it('should render as a <div> by default', () => {
  const wrapper = shallow(<Raw />);

  expect(wrapper.type()).toBe('div');
});

it('should render with a custom tag', () => {
  const wrapper = shallow(<Raw tag="span" />);

  expect(wrapper.type()).toBe('span');
});

describe('nameAs', () => {
  it('should apply the htmlFor attribute to <Label />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Label>Label</Label>
      </Raw>,
    );

    expect(wrapper.find(Label).props().htmlFor).toBe('foo');
  });

  it('should preserve custom htmlFor on <Label />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Label htmlFor="bar">Label</Label>
      </Raw>,
    );

    expect(wrapper.find(Label).props().htmlFor).toBe('bar');
  });

  it('should apply the id attribute to <Input />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Input />
      </Raw>,
    );

    expect(wrapper.find(Input).props().id).toBe('foo');
  });

  it('should preserve custom id on <Input />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Input id="bar" />
      </Raw>,
    );

    expect(wrapper.find(Input).props().id).toBe('bar');
  });

  it('should apply the name attribute to <Input />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Input />
      </Raw>,
    );

    expect(wrapper.find(Input).props().name).toBe('foo');
  });

  it('should preserve custom name on <Input />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Input name="bar" />
      </Raw>,
    );

    expect(wrapper.find(Input).props().name).toBe('bar');
  });

  it('should apply the id attribute to <Textarea />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Textarea />
      </Raw>,
    );

    expect(wrapper.find(Textarea).props().id).toBe('foo');
  });

  it('should preserve custom id on <Textarea />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Textarea id="bar" />
      </Raw>,
    );

    expect(wrapper.find(Textarea).props().id).toBe('bar');
  });

  it('should apply the name attribute to <Textarea />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Textarea />
      </Raw>,
    );

    expect(wrapper.find(Textarea).props().name).toBe('foo');
  });

  it('should preserve custom name on <Textarea />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Textarea name="bar" />
      </Raw>,
    );

    expect(wrapper.find(Textarea).props().name).toBe('bar');
  });

  //

  it('should apply the id attribute to <Select />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Select />
      </Raw>,
    );

    expect(wrapper.find(Select).props().id).toBe('foo');
  });

  it('should preserve custom id on <Select />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Select id="bar" />
      </Raw>,
    );

    expect(wrapper.find(Select).props().id).toBe('bar');
  });

  it('should apply the name attribute to <Select />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Select />
      </Raw>,
    );

    expect(wrapper.find(Select).props().name).toBe('foo');
  });

  it('should preserve custom name on <Select />', () => {
    const wrapper = shallow(
      <Raw nameAs="foo">
        <Select name="bar" />
      </Raw>,
    );

    expect(wrapper.find(Select).props().name).toBe('bar');
  });
});
