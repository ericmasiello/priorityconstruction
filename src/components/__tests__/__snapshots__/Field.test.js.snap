// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  margin-bottom: 1rem;
}

.c0 .sc-bdVaJa {
  -webkit-flex: 0 1 100px;
  -ms-flex: 0 1 100px;
  flex: 0 1 100px;
  max-width: 200px;
}

.c0 .sc-EHOje,
.c0 .sc-bZQynM,
.c0 .sc-htoDjs {
  width: 100%;
}

<div
  className="c0"
/>
`;

exports[`should render as stack 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 1rem;
}

.c0 .sc-EHOje,
.c0 .sc-bZQynM,
.c0 .sc-htoDjs {
  width: 100%;
}

<div
  className="c0"
/>
`;
